### ISSUE TITLE
Mix() can't find the mix-manifest.json for a Wordpress child theme

### ISSUE THREAD

**Issue by parsepixel** [mindkomm/theme-lib-mix#1]
Using `mix()` on a custom theme with laravel mix works great. But noticed it won't find the manifest if the $manifest_directory is in the child-theme.

In line 24 `get_theme_file_path`first checks `get_stylesheet_directory` which returns the path to the parent theme before `get_template_directory` which would be the child-theme.

Nobody should be modifying or adding anything to the parent theme and should compile all assets to the child, or would you recommend leaving the `mix-manifest.json` outside the theme entirely and pass the $manifest_directory every time?

**Comment by gchtr** [comment]
I looked at this more closely and I can definitely fix this.

The problem currently is that the `mix()` caches the manifest file the first time it is called and uses that manifest file for all calls that come next. Now if you use a child theme, it will still always use the parent theme’s files. That’s really not optimal and I can definitely improve this.

But even if I fix this, the problem is that if I make it work correctly with the `get_theme_file_path()` and `get_theme_file_url()` functions, it will always use the child theme manifest if it has the same relative path as the file in the parent theme. Even if you call it from the parent theme.

To get around that, I think we need a `mix_child()` function that will always return files from the child theme and use the parent files as a fallback.

> Nobody should be modifying or adding anything to the parent theme and should compile all assets to the child, or would you recommend leaving the `mix-manifest.json` outside the theme entirely and pass the $manifest_directory every time?

I want to make the package versatile and it should be up to you to choose the best strategy for your project. Passing the path to the manifest directory is not so nice, either. I’ll see whether I can improve that as well.

I’ll keep you posted about updates.

**Comment by audetcameron** [comment]
```php
function mix( $path, $manifest_directory = 'build', $child_theme = false ) {
	static $manifest;
	static $manifest_path;

	if ( ! $manifest_path ) {
		$manifest_path = get_theme_file_path( $manifest_directory . '/mix-manifest.json' );
	}

	// Bailout if manifest couldn’t be found
	if ( ! file_exists( $manifest_path ) ) {
		return get_theme_file_uri( $path );
	}

	if ( ! $manifest ) {
		// @codingStandardsIgnoreLine
		$manifest = json_decode( file_get_contents( $manifest_path ), true );
	}

	// Remove manifest directory from path
	$path = str_replace( $manifest_directory, '', $path );
	// Make sure there’s a leading slash
	$path = '/' . ltrim( $path, '/' );

	// Bailout with default theme path if file could not be found in manifest
	if ( ! array_key_exists( $path, $manifest ) ) {
		return get_theme_file_uri( $path );
	}

	// Get file URL from manifest file
	$path = $manifest[ $path ];
	// Make sure there’s no leading slash
	$path = ltrim( $path, '/' );

	if($child_theme){

		$file = ltrim( trailingslashit( $manifest_directory ) . $path, '/' );

		$url = get_stylesheet_directory_uri() . '/' . $file;

		return apply_filters( 'theme_file_uri', $url, $file );


	}
	return get_theme_file_uri( trailingslashit( $manifest_directory ) . $path );
}
```

This worked for me, I just pass a third param boolean if its a child theme, and manually apply the filter.

**Comment by parsepixel** [comment]
I had a very similar solution last year.

I removed the $manifest_directory argument all together as it became obsolete once I commented out the first bailout as `get_theme_file_path()` would always returning the path to the parent theme. I just used $is_child_theme argument to switch between using `get_theme_file_path` or `get_template_directory`

I'm glad I left my comments in place as I total forgot about this.

```php
function mix($path, $is_child_theme = false)
{
    static $manifest;
    static $manifest_path;

    // If child-theme this will always not be found get_theme_file_path() returns path to parent first
    // if ( ! $manifest_path ) {
    // 	$manifest_path = get_theme_file_path( $manifest_directory . '/mix-manifest.json' );
    // }

    //So this is broken and will always not find the file and do the bailout
    //Let check if we passed is child theme attribute
    if ($is_child_theme) {
        $manifest_path = get_theme_file_path('/mix-manifest.json');
    } else {
        $manifest_path = get_template_directory() . '/mix-manifest.json';
    }

    // Bailout if manifest couldn’t be found
    if (!file_exists($manifest_path)) {
        if ($is_child_theme) {
            return get_stylesheet_directory_uri() . '/' . $path;
        } else {
            return get_theme_file_uri($path);
        }
    }

    if (!$manifest) {
        // @codingStandardsIgnoreLine
        $manifest = json_decode(file_get_contents($manifest_path), true);
    }

    // Can skip this since we removed passing manifest directory for now
    // Remove manifest directory from path
    // $path = str_replace( $manifest_directory, '', $path );

    // Make sure there’s a leading slash
    $path = '/' . ltrim($path, '/');

    // Bailout with default theme path if file could not be found in manifest
    if (!array_key_exists($path, $manifest)) {
        if ($is_child_theme) {
            return get_stylesheet_directory_uri() . $path;
        } else {
            return get_theme_file_uri($path);
        }
    }

    // Get file URL from manifest file
    $path = $manifest[$path];
    // Make sure there’s no leading slash
    $path = ltrim($path, '/');

    if ($is_child_theme) {
        return trailingslashit(get_stylesheet_directory_uri()) . $path;
    } else {
        // commented out original since we removed directory
        // get_theme_file_uri( trailingslashit( $manifest_directory ) . $path );
        return get_theme_file_uri(trailingslashit($path));
    }
}
```

**Comment by parsepixel** [comment]
I found a more recent version I updated that kept the $manifest_directory and adjusted using `get_template_directory` for `get_theme_file_path` and `get_theme_file_uri` for `get_stylesheet_directory_uri` when the new $is_child_theme is true.

```php
function mix( $path, $manifest_directory = 'build', $is_child_theme = false) {
	static $manifest;
	static $manifest_path;

	// Used with child-theme get_theme_file_path() returns path to parent first
	if ( ! $manifest_path ) {
		$manifest_path = $is_child_theme ? get_theme_file_path( $manifest_directory . '/mix-manifest.json' ) : get_template_directory(). $manifest_directory . '/mix-manifest.json';
	}

	// Bailout if manifest couldn’t be found
	if ( ! file_exists( $manifest_path ) ) {
		return ($is_child_theme) ? get_stylesheet_directory_uri().$path : get_theme_file_uri($path);
	}

	if ( ! $manifest ) {
		// @codingStandardsIgnoreLine
		$manifest = json_decode( file_get_contents( $manifest_path ), true );
	}

	// Remove manifest directory from path
	if ( ! $is_child_theme) {
		$path = str_replace( $manifest_directory, '', $path );
	} else {
		$path = str_replace(get_stylesheet_directory_uri(), '', $path);
	}

	// Make sure there’s a leading slash
	$path = '/' . ltrim( $path, '/' );
	
	// Bailout with default theme path if file could not be found in manifest
	if ( ! array_key_exists( $path, $manifest ) ) {
		return ($is_child_theme) ? get_stylesheet_directory_uri().$path : get_theme_file_uri($path);
	}

	// Get file URL from manifest file
	$path = $manifest[ $path ];
	// Make sure there’s no leading slash
	$path = ltrim( $path, '/' );

	return ($is_child_theme) ? trailingslashit( get_stylesheet_directory_uri() ) . $path : get_theme_file_uri( trailingslashit( $manifest_directory ) . $path );
}
```

**Comment by gchtr** [comment]
This is now fixed in the [1.2.0 release](https://github.com/mindkomm/theme-lib-mix/releases/tag/1.2.0). Thanks for your patience.

You can use `mix_child()` or the `is_child` argument for the second parameter in `mix()`.

```php
wp_enqueue_style(
	'styles-child',
	mix_child( 'build/css/styles-child.css' ),
	[],
	null
);

// or

wp_enqueue_style(
	'styles-child',
	mix( 'build/css/styles-child.css', [
        'is_child' => true,
    ] ),
	[],
	null
);
```

The second `$args` parameter in `mix()` is backwards compatible with the `$manifest_directory` parameter. I didn’t want to introduce a third parameter for `is_child`, because then you would have to pass the `$manifest_directory` path in all function calls (well, at least for PHP versions < 8).
